src/ed25519.rs

```diff
@@ -0,0 +1,188 @@
+use std::str::FromStr;
+
+use crate::{error::DecodeError, convert::{encode, decode}, version};
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct PrivateKey(pub [u8; 32]);
+
+impl PrivateKey {
+    pub fn as_string(&self) -> String {
+        encode(version::PRIVATE_KEY_ED25519, &self.0)
+    }
+
+    pub fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(ed25519) => Ok(Self(ed25519)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::PRIVATE_KEY_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+impl FromStr for PrivateKey {
+    type Err = DecodeError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        PrivateKey::from_string(s)
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct PublicKey(pub [u8; 32]);
+
+impl PublicKey {
+    pub fn as_string(&self) -> String {
```

Comment:
`as_` functions need to return borrowed data, but this function is returning owned data. `to_` functions are the name that should be given to any data that is borrowed -> owned. See https://rust-lang.github.io/api-guidelines/naming.html.

If the goal is to avoid a conflict with a Display impl, a conflict should be fine? We can an inherent and a trait impl that does the same thing, and we can add the Display impl.

Reply:
Oh I see the `to_string` is generated from a blanket impl, so you can still write your own `to_string`. Will rename then!